soroban-env-host/src/vm/parsed_module.rs

```diff
@@ -153,32 +153,62 @@ impl ParsedModule {
         engine: &Engine,
         wasm: &[u8],
         cost_inputs: VersionedContractCodeCostInputs,
-    ) -> Result<Self, HostError> {
+    ) -> Result<Rc<Self>, HostError> {
         cost_inputs.charge_for_parsing(host)?;
         let (module, proto_version) = Self::parse_wasm(host, engine, wasm)?;
-        Ok(Self {
+        Ok(Rc::new(Self {
             module,
             proto_version,
             cost_inputs,
-        })
+        }))
     }
 
-    #[cfg(any(test, feature = "testutils"))]
+    pub fn with_import_symbols<T>(
+        &self,
+        callback: impl FnOnce(&BTreeSet<(&str, &str)>) -> Result<T, HostError>,
+    ) -> Result<T, HostError> {
+        // Cap symbols we're willing to import at 10 characters for each of
+        // module and function name. in practice they are all 1-2 chars, but
+        // we'll leave some future-proofing room here. The important point
+        // is to not be introducing a DoS vector.
+        const SYM_LEN_LIMIT: usize = 10;
+        let symbols: BTreeSet<(&str, &str)> = self
```

Comment:
Erring on extra caution, would it be plausible to use a `MeteredVec` here? 
There is a chance the contract can have lots of import entries (greater than number of available host functions), in which case the set construction and followup look ups costs are non-trivial.
Using `MeteredVec` also avoids dependency on the hash algorithm that isn't totally under our control.

Reply:
The costs are accounted for in the new cost model: it has a term specifically for the number of imports, which is calibrated to cover this cost.

There is no dependency on a hash algorithm here. BTreeSet is a b-tree, it does comparisons. I also bounded the length of import symbols at 10 characters, so there's no possibility of an over-long comparison.

Reply:
sounds good